server/core/src/atomic_apply_database_operations.cpp

```diff
@@ -0,0 +1,757 @@
+                std::ostringstream conditions_stream;
+                std::copy(std::begin(condition_clauses),
+                          std::end(condition_clauses),
+                          std::experimental::make_ostream_joiner(conditions_stream, " and "));
+
+                // Generate the final SQL.
+                const auto sql = fmt::format("update {} set {} where {}", table_name, set_stream.str(), conditions_stream.str());
+                rodsLog(LOG_NOTICE, "%s :: Generated SQL = [%s]", __func__, sql.data());
+
+#ifdef ATOMIC_DB_OPS_RUN_SQL
+                nanodbc::statement stmt{_db_conn};
+                nanodbc::prepare(stmt, sql);
+
+                // Bind condition values to prepared statement parameter markers.
+                for (std::size_t i = 0; i < bind_args.size(); ++i) {
+                    stmt.bind(i, bind_args[i].data());
+                }
+
+                nanodbc::execute(stmt);
+#endif
+            }
+        }
+        catch (const std::exception& e) {
+            THROW(SYS_LIBRARY_ERROR, e.what());
+        }
+
+        return 0;
+    } // update_rows
+
+    auto throw_if_invalid_table_column(const std::string_view _column,
+                                       const std::string_view _table_name) -> void
+    {
+        const auto iter = std::find(std::begin(db_columns),
+                                    std::end(db_columns),
+                                    db_column_type{_table_name, _column});
+
+        if (iter == std::end(db_columns)) {
+            const auto msg = fmt::format("Unsupported table column [{1}.{0}].", _column, _table_name);
+            rodsLog(LOG_ERROR, "%s :: %s", __func__, msg.data());
+            THROW(SYS_INVALID_INPUT_PARAM, msg);
+        }
+    } // throw_if_invalid_table_column
+
+    auto throw_if_invalid_conditional_operator(const std::string_view _operator) -> void
+    {
+        const auto iter = std::find(std::begin(supported_operators),
+                                    std::end(supported_operators),
+                                    _operator);
+
+        if (iter == std::end(supported_operators)) {
+            const auto msg = fmt::format("Unsupported conditional operator [{}].", _operator);
+            rodsLog(LOG_ERROR, "%s :: %s", __func__, msg.data());
+            THROW(SYS_INVALID_INPUT_PARAM, msg);
+        }
+    } // throw_if_invalid_conditional_operator
+} // anonymous namespace
+
+namespace irods::experimental
+{
+    auto atomic_apply_database_operations(RsComm& _comm, const nlohmann::json& _ops) -> int
+    {
+        rodsLog(LOG_NOTICE, "%s :: input = %s", __func__, _ops.dump(4).data());
+
+        std::string db_instance_name;
+        nanodbc::connection db_conn;
+        std::tie(db_instance_name, db_conn) = ic::new_database_connection();
+
+        // The "&dbc = db_conn" is required to avoid a bug with the C++ standard around structured bindings.
+        // See https://stackoverflow.com/a/46115028/5818611 for details.
```

Comment:
This comment can be removed if structured binding aren't used to extract `db_instance_name` and `db_conn`.